explorer/interpreter/interpreter.cpp

```diff
@@ -1087,7 +1085,23 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
       if (auto rewrite = op.rewritten_form()) {
         return todo_.ReplaceWith(std::make_unique<ExpressionAction>(*rewrite));
       }
-      if (act.pos() != static_cast<int>(op.arguments().size())) {
+      if ((op.op() == Operator::And || op.op() == Operator::Or) &&
+          act.pos() >= 1) {
+        auto operand_value =
+            cast<BoolValue>(*act.results()[act.pos() - 1]).value();
+        if ((op.op() == Operator::Or && operand_value) ||
+            (op.op() == Operator::And && !operand_value) || act.pos() == 2) {
+          // Short-circuit evaluation for 'and'/'or'
+          CARBON_ASSIGN_OR_RETURN(Nonnull<const Value*> value,
+                                  ErrorOr<Nonnull<const Value*>>(
+                                      arena_->New<BoolValue>(operand_value)));
+          return todo_.FinishAction(value);
```

Comment:
This can be simplified a little to reuse the same value we just produced:
```suggestion
        auto operand_value = cast<BoolValue>(act.results()[act.pos() - 1]);
        if ((op.op() == Operator::Or && operand_value->value()) ||
            (op.op() == Operator::And && !operand_value->value()) || act.pos() == 2) {
          return todo_.FinishAction(operand_value);
```

Reply:
Ha yes, much neater, didn't realize this was an arena value but it makes sense now 👍